distributed/scheduler.py

```diff
@@ -7491,19 +7491,16 @@ def decide_worker(
     if ts._actor:
         candidates = set(all_workers)
     else:
+        # Select all workers holding deps of this task
         candidates = {wws for dts in deps for wws in dts._who_has}
-        # Add some random workers to into `candidates`, starting with idle ones
-        # TODO shuffle to prevent hotspots?
-        N_RANDOM_WORKERS: Py_ssize_t = 20
-        candidates.update(idle_workers[:N_RANDOM_WORKERS])
-        if len(idle_workers) < N_RANDOM_WORKERS:
-            sample_from = (
-                list(valid_workers) if valid_workers is not None else all_workers
-            )
-            candidates.update(
-                random.sample(sample_from, min(N_RANDOM_WORKERS, len(sample_from)))
-                # ^ NOTE: `min` because `random.sample` errors if `len(sample) < k`
-            )
+        # Add up to 10 random workers into `candidates`, preferring idle ones.
+        sample_from = (
+            list(valid_workers)
+            if valid_workers is not None
+            else idle_workers or all_workers
+        )
+        candidates.update(random.sample(sample_from, min(10, len(sample_from))))
+        # ^ NOTE: `min` because `random.sample` errors if `len(sample) < k`
```

Comment:
```python
import random
import sortedcontainers

d = sortedcontainers.SortedDict()
for i in range(100):
    d[str(i)] = i

candidates = set()

%timeit candidates.update(random.sample(d.values(), min(10, len(d))))
```
```
12.4 µs ± 652 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

You're not wrong to be concerned.  12us is a decent chunk of time.  

It looks like `random.sample` isn't cheap

```python
%timeit random.sample(d.values(), min(10, len(d)))
# 11.7 µs ± 61.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

%timeit d.values()[:10]
# 2.09 µs ± 27.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Reply:
This might help with efficiency while also up with simplicity?

```python
        candidates = {wws for dts in deps for wws in dts._who_has}

        extra = (
            list(valid_workers)
            if valid_workers is not None
            else (idle_workers or all_workers)
        )
        if len(extra) > 10:
            extra = random.sample(extra, 10)
        
        candidates.update(extra)
```

Reply:
Some time is spent just working with `sortedcontainers`.

```python
import random

d = dict()
for i in range(100):
    d[str(i)] = i

candidates = set()

%timeit candidates.update(random.sample(list(d.values()), min(10, len(d))))
```

```
7.21 µs ± 31.3 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

As a side note it seems `random.sample` won't take a `dict` or its iterators directly. So this includes coercing to a `list`

Reply:
```python
l = list(d.values())

%timeit candidates.update(random.sample(l, min(10, len(l))))
# 5.09 µs ± 179 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

%timeit l = list(d.values()[:10])
1.98 µs ± 34.9 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

%timeit l[:10]
# 79.3 ns ± 2.2 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

SortedValuesView [does say](https://github.com/grantjenks/python-sortedcontainers/blob/master/sortedcontainers/sorteddict.py#L781) lookups are O(log(n)), but it seems like it's more than that.

Reply:
It takes any indexable, which `sortedcontainers.SortedDict.values` provides.  Indexability is the reason why we use sortedcontainers.

Reply:
Right the point is that results in it taking ~2x longer than just copying to `list` and feeding it in

Reply:
Oh interesting

Reply:
```python
        candidates = {wws for dts in deps for wws in dts._who_has}

        if valid_workers is not None:
            extra = valid_workers
        elif idle_workers:
            extra = idle_workers[:10]
        else:
            extra = all_workers

        if len(extra) > 10:
            extra = random.sample(extra, 10)
        
        candidates.update(extra)
```

This might be getting too deep into the weeds though.

Reply:
For comparison:
```python
from distributed.scheduler import TaskState, WorkerState

N = 5
extra = 2
workers = [WorkerState(str(i)) for i in range(N)]
deps = [TaskState(str(i), None) for i in range(N)]
for dts, ws in zip(deps, workers):
    dts.who_has.add(ws)

for i in range(extra):
    deps[i].who_has.add(workers[-i])
deps = set(deps)

%timeit candidates = {wws for dts in deps for wws in dts._who_has}
971 ns ± 24 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
```
N = 50
extra = 50
12 µs ± 202 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

N = 100
extra = 10
17.2 µs ± 214 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

So just to be clear that the random sample is the thing to try to avoid, not the `who_has` selection. In the uncommon case of many many dependencies duplicated over many workers, that selection is still on par with the random sample.

Reply:
So sampling the keys and then using them to look up values is 2x faster with SortedDict, and the same as what we'd get with a plain dict:
```python
%timeit candidates.update(d[k] for k in random.sample(list(d), 10))
6.95 µs ± 227 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

pd = dict(d)
%timeit candidates.update(pd[k] for k in random.sample(list(pd), 10))
6.36 µs ± 95.5 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Also `random.choices` is 2x faster than `random.sample` (makes sense since it doesn't have to hold onto any state):
```python
%timeit candidates.update(d[k] for k in random.choices(list(d), k=10))
3.65 µs ± 11.9 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

%timeit candidates.update(pd[k] for k in random.choices(list(pd), k=10))
3.17 µs ± 36.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

I'd feel okay about `random.choices`; we'd get slightly fewer extra workers in the pool, but it's probably worth the time savings.

Reply:
If we're okay with duplicates then we can go much faster.

Iterating values of a SortedDict is slow:
```python
%timeit list(d.values())
8.43 µs ± 85.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

But it's fast on a plain dict:
```python
pd = dict(d)
%timeit list(pd.values())
691 ns ± 5.07 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
So we just have to make sure we're passing in `self._workers_dv.values()`, not `self._workers.values()`.

This is decent:
```python
%timeit candidates.update(random.choices(list(pd.values()), k=10))
2.45 µs ± 15.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

But we can go much faster and avoid a copy with an iterator:

```python
import itertools
import random

def random_choices_iter(xs: "Iterator", k: int) -> "Iterator":
    return itertools.islice(itertools.takewhile(lambda _: random.random() < 0.5, xs), k)

%timeit candidates.update(random_choices_iter(pd.values(), k=10))
643 ns ± 12.5 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Reply:
It looks `random.sample` will take `d.keys()` from playing around with it locally. So we can probably skip the coercion to a `list` in that case. Though maybe we've already moved well beyond this at this point

I'd probably shy away from getting to clever around random item selection. It's probably more likely to have surprising behavior that we only discover later. Just my 2 cents 🙂

Reply:
I think we're beyond that, since we're only taking `dict_values` objects in this function now. Making it able to support both dicts and sets was a bit of a mess and not worth it IMO.

I didn't go with the too-clever itertools option because it did have surprising behavior. I did have to add another conditional though because `choices` can repeat items: fcb165e. So perhaps even that's too clever?